### ISSUE TITLE
HDF5 1.12 Bindings Broken

### ISSUE THREAD

**Issue by NAThompson** [ornladios/ADIOS2#2083]
Using HDF5 installed on Mac:

```
$ brew install hdf5
==> Installing hdf5
==> Downloading https://homebrew.bintray.com/bottles/hdf5-1.12.0.mojave.bottle.tar.gz
==> Downloading from https://akamai.bintray.com/77/77cf2db15c52c69da64ca5535eb03e44ec2bd953cbb2166996580739db467241?__gda__=ex
######################################################################## 100.0%
==> Pouring hdf5-1.12.0.mojave.bottle.tar.gz
```

We have the following error:

```
ADIOS2/build$ ninja
-- Build files have been written to: /Users/4nt/ADIOS2/build
[7/373] Building CXX object source/adios2/CMakeFiles/adios2_hdf5.dir/toolkit/interop/hdf5/HDF5Common.cpp.o
FAILED: source/adios2/CMakeFiles/adios2_hdf5.dir/toolkit/interop/hdf5/HDF5Common.cpp.o 
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/c++  -DMPICH_SKIP_MPICXX -DOMPI_SKIP_MPICXX -I../source -Isource -I/usr/local/Cellar/hdf5/1.12.0/include -I/usr/local/opt/szip/include -O3 -DNDEBUG -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.15.sdk -mmacosx-version-min=10.14 -fPIC   -std=gnu++11 -MD -MT source/adios2/CMakeFiles/adios2_hdf5.dir/toolkit/interop/hdf5/HDF5Common.cpp.o -MF source/adios2/CMakeFiles/adios2_hdf5.dir/toolkit/interop/hdf5/HDF5Common.cpp.o.d -o source/adios2/CMakeFiles/adios2_hdf5.dir/toolkit/interop/hdf5/HDF5Common.cpp.o -c ../source/adios2/toolkit/interop/hdf5/HDF5Common.cpp
../source/adios2/toolkit/interop/hdf5/HDF5Common.cpp:1263:18: error: no matching function for call to 'H5Oget_info3'
    herr_t ret = H5Oget_info(m_FileId, &oinfo);
                 ^~~~~~~~~~~
/usr/local/Cellar/hdf5/1.12.0/include/H5version.h:746:23: note: expanded from macro 'H5Oget_info'
  #define H5Oget_info H5Oget_info3
                      ^~~~~~~~~~~~
/usr/local/Cellar/hdf5/1.12.0/include/H5Opublic.h:187:15: note: candidate function not viable: requires 3 arguments, but 2 were provided
H5_DLL herr_t H5Oget_info3(hid_t loc_id, H5O_info2_t *oinfo, unsigned fields);
              ^
```

**Comment by williamfgc** [comment]
Looks more like `HDF5` is [breaking its API](https://www.hdfgroup.org/2020/03/release-of-hdf5-1-12-0-newsletter-172/). Version 1.12 is fairly recent

**Comment by NAThompson** [comment]
Yup, they have.

I also tried to use their suggestion to add the `CMake` flag `–DH5_USE_110_API:BOOL=ON`; no joy.

**Comment by lwan86** [comment]
@NAThompson When you brew install, have you enabled the parallel hdf5?

**Comment by NAThompson** [comment]
@lwan86 : No, it's vanilla.

**Comment by lwan86** [comment]
@NAThompson If I remember correctly, ADIOS doesn't compile if it's not parallel hdf5. Maybe I'm wrong, but please double check.

**Comment by NAThompson** [comment]
No joy:

```
ADIOS2/build$ brew uninstall hdf5
ADIOS2/build$ brew install iltommi/brews/hdf5-parallel
ADIOS2/build$ brew link hdf5-parallel --force
ADIOS2/build$ cmake ../ -G Ninja
ADIOS2/build$ ninja
../source/adios2/toolkit/interop/hdf5/HDF5Common.cpp:1263:18: error: no matching function for call to 'H5Oget_info3'
    herr_t ret = H5Oget_info(m_FileId, &oinfo);
                 ^~~~~~~~~~~
/usr/local/include/H5version.h:746:23: note: expanded from macro 'H5Oget_info'
  #define H5Oget_info H5Oget_info3
                      ^~~~~~~~~~~~
/usr/local/include/H5Opublic.h:187:15: note: candidate function not viable: requires 3 arguments, but 2 were provided
H5_DLL herr_t H5Oget_info3(hid_t loc_id, H5O_info2_t *oinfo, unsigned fields);
              ^
../source/adios2/toolkit/interop/hdf5/HDF5Common.cpp:1314:18: error: no matching function for call to 'H5Oget_info3'
    herr_t ret = H5Oget_info(datasetId, &oinfo);
                 ^~~~~~~~~~~
/usr/local/include/H5version.h:746:23: note: expanded from macro 'H5Oget_info'
  #define H5Oget_info H5Oget_info3
                      ^~~~~~~~~~~~
/usr/local/include/H5Opublic.h:187:15: note: candidate function not viable: requires 3 arguments, but 2 were provided
H5_DLL herr_t H5Oget_info3(hid_t loc_id, H5O_info2_t *oinfo, unsigned fields);
              ^
2 errors generated.
```

**Comment by chuckatkins** [comment]
I actually just spoke with HDF group about this earlier today.  They have an API breakage from 1.10 to 1.12.  However, 1.12 can be built with the 1.10 compatibility APIs.

**Comment by chuckatkins** [comment]
Is this sufficient?

```cmake
diff --git a/source/adios2/CMakeLists.txt b/source/adios2/CMakeLists.txt
index 8dba9c6d4..68e846bc1 100644
--- a/source/adios2/CMakeLists.txt
+++ b/source/adios2/CMakeLists.txt
@@ -269,6 +269,9 @@ if(ADIOS2_HAVE_HDF5)
 
   foreach(lib IN LISTS adios2_hdf5_objs)
     target_link_libraries(${lib} PRIVATE ${HDF5_C_LIBRARIES})
+    if(HDF5_VERSION VERSION_GREATER_EQUAL 1.10.0)
+      target_compile_definitions(${lib} PRIVATE H5_USE_110_API)
+    endif()
     if(HDF5_IS_PARALLEL)
       target_link_libraries(${lib} PRIVATE $<BUILD_INTERFACE:MPI::MPI_C>)
     endif()
```

**Comment by germasch** [comment]
I'd say > instead of >= in that test should be sufficient. (or >= 1.12.0).
That's the fix that I've seen other packages use, but I haven't verified it for adios2.

**Comment by guj** [comment]
Yes, that’s right. - Junmin

**Comment by NAThompson** [comment]
I have successfully produced an HDF5 file whose data appears to be correct with this fix.

**Comment by chuckatkins** [comment]
Fixed by #2091
